### PR TITLE
Custom types support

### PR DESCRIPTION
**Pull request by alexpaschenko**

### CODE REVIEW

File: h2/src/main/org/h2/api/CustomType.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+package org.h2.api;
+
+import org.h2.store.DataHandler;
+import org.h2.value.Value;
+
+import java.util.ArrayList;
+
+/**
+ * Interface for custom types to define their construction logic
+ * based on more 'primitive' {@link Value}s.
```
Comment (svladykin) [inline h2/src/main/org/h2/api/CustomType.java]: There are no more primitive or less primitive types. Just say that it is a conversion logic from builtin types.

File: h2/src/main/org/h2/api/ErrorCode.java [hunk 2]
```diff
@@ -1935,7 +1935,29 @@
...
+     * The error with code <code>90144</code> is thrown when
+     * trying to register custom data type if an object with this name already exists.
+     * Example:
+     *+
+     * CREATE VALUE TYPE FOO FOR "org.acme.Foo";
+     * CREATE VALUE TYPE BAR FOR "org.acme.Bar";
+     * CREATE VALUE TYPE BAR FOR "org.acme.Bar";
+     * </pre>
+     */
+    public static final int CUSTOM_DATA_TYPE_ALREADY_EXISTS_1 = 90144;
```
Comment (svladykin) [inline h2/src/main/org/h2/api/ErrorCode.java]: Please provide meaningful examples for types. For example H2 does not have builtin support for complex numbers or matrices.

File: h2/src/main/org/h2/command/Parser.java [hunk 3]
```diff
@@ -1527,6 +1532,16 @@ private DropUserDataType parseDropUserDataType() {
         return command;
     }
 
+    private DropCustomDataType parseDropCustomDataType() {
```
Comment (svladykin) [inline h2/src/main/org/h2/command/Parser.java]: Please replace the word `custom` with `value` everywhere.

File: h2/src/main/org/h2/command/Parser.java [hunk 4]
```diff
@@ -4138,6 +4154,8 @@ private Column parseColumnWithType(String columnName) {
             precision = templateColumn.getPrecision();
             displaySize = templateColumn.getDisplaySize();
             scale = templateColumn.getScale();
+        } else if (customDataType != null) {
+            dataType = DataType.getTypeByName("JAVA_OBJECT");
```
Comment (svladykin) [inline h2/src/main/org/h2/command/Parser.java:4158]: This stuff looks suspicious

File: h2/src/main/org/h2/command/ddl/CreateCustomDataType.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import org.h2.table.Table;
+import org.h2.util.JdbcUtils;
+import org.h2.value.DataType;
+
+/**
+ * This class represents the statement CREATE VALUE TYPE
+ *
+ * @author apaschenko
+ */
+public class CreateCustomDataType extends DefineCommand {
```
Comment (svladykin) [inline h2/src/main/org/h2/command/ddl/CreateCustomDataType.java]: `CreateValueType`

File: h2/src/main/org/h2/expression/Comparison.java [hunk 6]
```diff
@@ -200,7 +203,7 @@ public Expression optimize(Session session) {
                     // If not the column values will need to be promoted
                     // to constant type, but vise versa, then let's do this here once.
                     if (constType != resType)
-                        right = ValueExpression.get(r.convertTo(resType));
+                        right = ValueExpression.get(((ExpressionColumn) left).getColumn().convert(r));
```
Comment (svladykin) [inline h2/src/main/org/h2/expression/Comparison.java:204]: This change looks dirty as well.

File: h2/src/test/org/h2/test/testScript.sql [hunk 7]
```diff
@@ -3196,6 +3196,18 @@ drop domain email;
...
+> CLASS_NAME                            SQL
+> ------------------------------------- ----------------------------------------------------------------------------------------
+> org.h2.samples.CustomTypes$PersonType CREATE VALUE TYPE PERSON FOR "org.h2.samples.CustomTypes$PersonType" WITH PARAM1, PARAM2
+> rows: 1
+
+drop VALUE TYPE person;
+> ok
+
 create force view address_view as select * from address;
 > ok
```
Comment (svladykin) [inline h2/src/test/org/h2/test/testScript.sql:3213]: Overall the binding to a column looks wrong. The following stuff have to work without any columns:
```
SELECT CAST ('1 + 3i' AS COMPLEX) + '2 + 2i'
```

**Review by svladykin — CHANGES_REQUESTED** [review]

File: h2/src/main/org/h2/command/Parser.java (hunk 4 above)
Comment (alexpaschenko) [reply h2/src/main/org/h2/command/Parser.java:4158]: Probably we should let implementations declare what type they boil down to, what do you think? Here `JAVA_OBJECT` is present simply as long as it is the most generic thing possible.

File: h2/src/main/org/h2/expression/Comparison.java (hunk 6 above)
Comment (alexpaschenko) [reply h2/src/main/org/h2/expression/Comparison.java:204]: I don't really see why. Before this change, `Column`'s `convert` did exactly the same thing as the old  conversion code here (and could have been used right away). Please see my thoughts about binding of the value type to column below. After all, here we clearly want to convert constant to type of the column, so asking the column to do the conversion seems rational.

File: h2/src/test/org/h2/test/testScript.sql (hunk 7 above)
Comment (alexpaschenko) [reply h2/src/test/org/h2/test/testScript.sql:3213]: The thing is, this cast (the part afrer `AS`, to be precise) is parsed exactly like any column in `CREATE TABLE` and actually even a `Column` object is created for a cast under the hood, so to me it is looks legit. Column knows what custom type is associated with it and can perform needed conversions transparently, without need for anyone from outside to bother about any new types. Do you have more concrete arguments why this binding is bad?

Comment (svladykin) [reply h2/src/test/org/h2/test/testScript.sql:3213]: Can you please add a test case like this?

File: h2/src/main/org/h2/command/ddl/CreateCustomDataType.java (hunk 5 above)
Comment (alexpaschenko) [reply h2/src/main/org/h2/command/ddl/CreateCustomDataType.java]: Fixed to `CreateUserValueType`

File: h2/src/main/org/h2/command/Parser.java (hunk 3 above)
Comment (alexpaschenko) [reply h2/src/main/org/h2/command/Parser.java]: Fixed

File: h2/src/main/org/h2/api/CustomType.java (hunk 1 above)
Comment (alexpaschenko) [reply h2/src/main/org/h2/api/CustomType.java]: Fixed

File: h2/src/main/org/h2/api/ErrorCode.java (hunk 2 above)
Comment (alexpaschenko) [reply h2/src/main/org/h2/api/ErrorCode.java]: Fixed

**Comment by svladykin** [comment]
Ok, here is an example of what will not work with your column binding:
```sql
select complex(t.r, t.i) + '2 + 2i' from t
```
where `complex` returns user defined `Value` of type `COMPLEX`.

This is why your approach is broken.

**Comment by alexpaschenko** [comment]
Erm. I'm sorry but why + operation should work on _objects_ in the first place? It seems to me that this thing is primarily for convenient comparisons and conversions on INSERT and not for trying to support quasi random stuff like adding one object to another.

**Pull request closed without merging**